### PR TITLE
etcdctl: add migrate command

### PR DESCRIPTION
**Pull request by xiang90**
/cc @heyitsanthony @hongchaodeng 

Migrate command accepts a datadir and an optional user-provided
transformer function that transform v2 keys to v2 keys.

Migrate command then builds a v3 backend state based on the existing
v2 keys and the output of the transformer function.

### CODE REVIEW

File: etcdctl/ctlv3/command/migrate_command.go [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+	"github.com/spf13/cobra"
+)
+
+var (
+	migrateDatadir     string
+	migrateWALdir      string
+	migrateTransformer string
+)
+
+// NewMigrate returns the cobra command for "Migrate".
```
Comment (gyuho) [inline etcdctl/ctlv3/command/migrate_command.go]: `// NewMigrateCommand ...`?

File: etcdserver/apply_v2.go [hunk 2]
```diff
@@ -35,62 +35,69 @@ type applierV2 interface {
 	Sync(r *pb.Request) Response
 }
 
-type applierV2store struct{ s *EtcdServer }
+type ApplierV2store struct {
```
Comment (heyitsanthony) [inline etcdserver/apply_v2.go]: can this stay private and instead have `func NewApplierV2Store(s *EtcdServer, c *membership.RaftCluster) ApplierV2`?

Comment (xiang90) [inline etcdserver/apply_v2.go]: ok. will try.

File: etcdctl/ctlv3/command/migrate_command.go [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+	toTTLOptions(r)
+	switch r.Method {
+	case "POST":
+		applyV2.Post(r)
+	case "PUT":
+		applyV2.Put(r)
+	case "DELETE":
+		applyV2.Delete(r)
+	case "QGET":
+		applyV2.QGet(r)
```
Comment (heyitsanthony) [inline etcdctl/ctlv3/command/migrate_command.go]: default: panic?

File: etcdctl/ctlv3/command/migrate_command.go [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+		ExitWithError(ExitError, err)
+	}
+	return writeKeys(w, all.Node)
+}
+
+func writeKeys(w io.Writer, n *store.NodeExtern) uint64 {
+	maxIndex := n.ModifiedIndex
+
+	nodes := n.Nodes
+	n.Key = n.Key[2:]
```
Comment (heyitsanthony) [inline etcdctl/ctlv3/command/migrate_command.go]: what's this removing?

Comment (xiang90) [inline etcdctl/ctlv3/command/migrate_command.go]: oh. i should comment this. the key space in v2 has a /1 prefix.

**Comment by gyuho** [comment]
Can we have some migration guide README after this?

File: etcdctl/ctlv3/command/migrate_command.go [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+	return make([]byte, revBytesLen)
+}
+
+func revToBytes(rev revision, bytes []byte) {
+	binary.BigEndian.PutUint64(bytes, uint64(rev.main))
+	bytes[8] = '_'
+	binary.BigEndian.PutUint64(bytes[9:], uint64(rev.sub))
+}
+
+func writeIndex(be backend.Backend, index uint64) {
```
Comment (heyitsanthony) [inline etcdctl/ctlv3/command/migrate_command.go]: this and writeToBackend seem like they belong in the mvcc package

Comment (xiang90) [inline etcdctl/ctlv3/command/migrate_command.go]: @heyitsanthony I will make that a public func in mvcc pkg.

File: etcdctl/ctlv3/command/migrate_command.go [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+			if _, err := dw.Write(data); err != nil {
+				panic(err)
+			}
+		}
+	}()
+
+	return sw, dr
+}
+
+func transform(n *client.Node) *mvccpb.KeyValue {
```
Comment (heyitsanthony) [inline etcdctl/ctlv3/command/migrate_command.go]: should this transform ttl'd keys into permanent keys?

Comment (xiang90) [inline etcdctl/ctlv3/command/migrate_command.go]: yes... this is what we do by default. just a very simple example transform func.

File: etcdctl/ctlv3/command/migrate_command.go (hunk 3 above)
Comment (xiang90) [inline etcdctl/ctlv3/command/migrate_command.go]: fixed

File: etcdctl/ctlv3/command/migrate_command.go [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+
+	errc := make(chan error, 1)
+
+	go func() {
+		defer func() {
+			sr.Close()
+			dw.Close()
+		}()
+
+		for {
```
Comment (heyitsanthony) [inline etcdctl/ctlv3/command/migrate_command.go]: `for decoder.More() {`

**Comment by heyitsanthony** [comment]
lgtm

**Pull request merged**